lib/formats/csv.flow

```diff
@@ -3,23 +3,125 @@ import ds/array;
 import math/math;
 import algorithms;
 import devmode;
+import lingo/pegcode/driver;
 
 export {
 	makeCsvFile(rows : [[string]]) -> string;
 	makeCsvFileSep(rows : [[string]], sep: string) -> string;
 
+	// NOTE: cannot correctly parse data that contains multiline strings
 	// If onlyRectangular is set, we return [[]] when there is not the same
 	// number of columns in each line. If it is not set, we can return an
 	// uneven array.
 	// This tries to detect the separator automatically
 	parseCsvFile(content : string, onlyRectangular : bool) -> [[string]];
 
+	// NOTE: cannot correctly parse data that contains multiline strings
 	// If you know the separator, use this one.
 	parseCsv(s : string, sep : string) -> [[string]];
 
 	parseCsvLine(s : string, sep : string) -> [string];
 
 	parseCsvField(s : string, index : int, sep : string) -> Pair<string, int>;
+
+	ParseCsvOption ::= ParseCsvSeparator, ParseCsvDetectSeparator, ParseCsvRectangle, ParseCsvSkipEmpty;
+		ParseCsvSeparator(sep : string); // any character except double quotes or empty string
+		ParseCsvDetectSeparator(); // try to detect if separator is either of , ; \t 
+		ParseCsvRectangle(); // treat inconsistent number of columns as an error
+		ParseCsvSkipEmpty(); // ignore empty rows, can be used with ParseCsvRectangle
+
+	// By default uses ',' as a separator.
+	// If an error occurs the second element of the pair will contain an error message
+	// and first element of the pair may contain a partial result.
+	parseCsvData(data : string, styles : [ParseCsvOption]) -> Pair<[[string]], string>;
+}
+
+compiledGrammarsR : ref Tree<string, [PegOp]> = ref makeTree();
+csvGrammarActions = SemanticActions(setTree(defaultPegActions.t, "unescapeQ", \args : [flow] -> {
+	str = flow2s(args[0]);
+	trimmed = substring(str, 1, strlen(str) - 2);
+	strReplace(trimmed, "\"\"", "\"");
+}));
+
+CsvRowInfo(start : int, end : int, row : [string]);
+
+parseCsvData(data : string, styles : [ParseCsvOption]) -> Pair<[[string]], string> {
+	styleSepM = tryExtractStruct(styles, ParseCsvSeparator(""));
+	sepStr = eitherFn(styleSepM, \styleSep -> styleSep.sep, \ -> {
+		if (containsStruct(styles, ParseCsvDetectSeparator())) {
+			sep = detectCsvSeparator(data).second;
+			devtrace("parseCsvData: detected separator is " + sep);
+			sep;
+		} else {
+			","; // default separator
+		}
+	});
+
+	// convert string to unicode escapes
+	sep = concatStrings(map(s2a(sepStr), \code -> {
+		hex = formatHex(code);
+		substring("\\u0000", 0, 6 - strlen(hex)) + hex;
+	}));
+
+	// cannot have double quotes or empty string as a separator
+	if (strContains(sep, "\\u0022") || sepStr == "") {
+		Pair([], "invalid separator");
+	} else {
+		grammarM = lookupTree(^compiledGrammarsR, sep);
+		grammar = eitherFn(grammarM, idfn, \ -> {
+			newOps = compilePegGrammar(formatString("#include formats/csv/csv.lingo-template", [sep]));
```

Comment:
the standard is *.lingo file extension

Reply:
Its not a working grammar before string formatting is applied